### ISSUE TITLE
EBUSY on overlayfs with recent linux updates

### ISSUE THREAD

**Issue by kmjohansen** [moby/moby#39475]
**Description**

After upgrading to linux 4.19.56 and docker 18.09.7, I'm seeing intermittent container creation failures due to problems with overlayfs.  This appears to be code that was recently added to Linux 5.2, and that has been subsequently backported to releases on the stable train.  The patch for my release is here:

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v4.19.56&id=f1c5aa5eda08710c2ba619d93126380881fa1114

This manifests as an error at container creation time:

```
Handler for POST /containers/create returned error: error creating overlay mount to /var/lib/docker/165536.165536/overlay2/e6ff2b345027c75a472003f6bd0ea84ea8f9a5d962fb48646ab22d06d06016c9/merged: device or resource busy
```

Looking at the `dmesg` for the corresponding time period, I can see:
```
overlayfs: lowerdir is in-use as upperdir/workdir
```
So we're pretty clearly running into the `ovl_is_inuse(stack[i].dentry)` check in `ovl_get_lower_layers()`.

A hypothesis that I haven't been able to yet prove is that the `-init` layer is getting captured in a child mountns as part of `unshare(2)` and that's what's causing us to hit the lower layer check.  It would seem that this has always been a possibility; however, linux has never policed the overlayfs layers to this degree until now.

Just looking at `strace(1)`, we can see a container create mounts a layer with an `-init` suffix as an upper layer before unmounting it, and then re-mounting a new overlay filesystem with that layer in the lower layers.  If this first mount is getting caught up in another mount namespace, that would lead to the situation where overlayfs believes the layer to still be in-use despite it being apparently unmounted in the host mount namespace.

As an example, here's the strace of mount/unmount for a `docker run -it --rm alpine`:

```
[pid 16589] mount("overlay", "/var/lib/docker/165536.165536/overlay2/ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a-init/merged", "overlay", 0, "index=off,lowerdir=/var/lib/docker/165536.165536/overlay2/l/7STKA6YTQ6NQQNXBRHUTZU6BM4,upperdir=/var/lib/docker/165536.165536/overlay2/ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a-init/diff,workdir=/var/lib/docker/165536.165536/overlay2/ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a-init/work") = 0
[pid 16589] umount2("/var/lib/docker/165536.165536/overlay2/ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a-init/merged", MNT_DETACH) = 0
[pid 16589] mount("overlay", "/var/lib/docker/165536.165536/overlay2/ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a/merged", "overlay", 0, "index=off,lowerdir=/var/lib/docker/165536.165536/overlay2/l/W4ABDZK7TSYI3PWOX3JZPAPMIB:/var/lib/docker/165536.165536/overlay2/l/7STKA6YTQ6NQQNXBRHUTZU6BM4,upperdir=/var/lib/docker/165536.165536/overlay2/ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a/diff,workdir=/var/lib/docker/165536.165536/overlay2/ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a/work") = 0
```

The `ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a-init/merged` filesystem mounts `/var/lib/docker/165536.165536/overlay2/ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a-init/diff` as its upper layer and then `umount`s the filesystem.  The subsequent mount of `ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a/merged` refers to `l/W4ABDZK7TSYI3PWOX3JZPAPMIB` which is itself a reference to `ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a-init/diff` see:

```
# ls -l l/W4ABDZK7TSYI3PWOX3JZPAPMIB
lrwxrwxrwx 1 root root 77 Jul  9 06:06 l/W4ABDZK7TSYI3PWOX3JZPAPMIB -> ../ebc3dbdd298ac565e605eed2f12d1e1063435152853d682524dea4e971e8af7a-init/diff
```

So it would seem that recent overlayfs changes have exposed a mountpoint leak that has existed in docker for some time, but that was effectively innocuous because of the relatively lax condition checking.  With the stricter checks in place, the mount is returning unsuccessfully with `EBUSY`.

**Steps to reproduce the issue:**
The issue has occurred multiple times within the past few days in multiple environments; however, I've been unable to work out the exact steps necessary for a deterministic reproduction.

**Describe the results you received:**
Got an error like:
```
Handler for POST /containers/create returned error: error creating overlay mount to /var/lib/docker/165536.165536/overlay2/e6ff2b345027c75a472003f6bd0ea84ea8f9a5d962fb48646ab22d06d06016c9/merged: device or resource busy
```

**Describe the results you expected:**

Expected the container to launch.

**Additional information you deem important (e.g. issue happens only occasionally):**

Issue happens infrequently, and has been difficult to reproduce so far.

I'm not the first person to encounter this issue.  I was able to turn up a couple of reports in other projects:

https://github.com/docker/for-linux/issues/711
https://github.com/cilium/packer-ci-build/pull/163

For context, it was very helpful to refer to a similar prior issue documented here:

https://github.com/moby/moby/issues/34672

This is not the same issue, but does go over some of the related problems.

**Output of `docker version`:**

```
# docker version
Client: Docker Engine - Community
 Version:           18.09.7
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        2d0083d
 Built:             Thu Jun 27 17:54:15 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.7
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       2d0083d
  Built:            Thu Jun 27 18:01:17 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
# docker info
Containers: 4
 Running: 1
 Paused: 0
 Stopped: 3
Images: 15
Server Version: 18.09.7
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan netplugin null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
init version: fec3683
Security Options:
 seccomp
  Profile: default
 userns
Kernel Version: 4.19.56-linuxkit-dbg
Operating System: Alpine Linux v3.10 (containerized)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 58.13GiB
Name: runner-020017080f0a
ID: GW7O:QTPA:DRLG:BRSU:MEIZ:XQ5Q:4HOO:L3A4:XGFW:GCP5:NQS4:KTFH
Docker Root Dir: /var/lib/docker/165536.165536
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Running on qemu/kvm based vms on intel x86-64 hardware.

**Comment by kmjohansen** [comment]
It looks like @amir73il is working on a fix for this.  I've commented on his patch, letting him know about a few of the related issues that have run into this: https://github.com/amir73il/linux/commit/9995d3a5ee3851a6883b0627b703e64146291453#commitcomment-34287909

**Comment by vladimiroff** [comment]
A [fix](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=0be0bfd2de9dfdd2098a9c5b14bdd8f739c9165d) for an overlay2 regression has just been released in 4.19.75, 5.2.17 and 5.3.1 which I believe should solve this.
